docs/architecture/adr-038-state-listening.md

```diff
@@ -204,60 +204,158 @@ func (rs *Store) CacheMultiStore() types.CacheMultiStore {
+	// ListenStoreKVPair updates the steaming service with the latest StoreKVPair messages
+	ListenStoreKVPair(blockHeight int64, data []byte) error
 }
 
-// StreamingService interface for registering WriteListeners with the BaseApp and updating the service with the ABCI messages using the hooks
-type StreamingService interface {
-    // Stream is the streaming service loop, awaits kv pairs and writes them to a destination stream or file
-    Stream(wg *sync.WaitGroup) error
-    // Listeners returns the streaming service's listeners for the BaseApp to register
-    Listeners() map[types.StoreKey][]store.WriteListener
-    // ABCIListener interface for hooking into the ABCI messages from inside the BaseApp
-    ABCIListener
-    // Closer interface
-    io.Closer
+// StreamingService struct for registering WriteListeners with the BaseApp and updating the service with the ABCI messages using the hooks
+type StreamingService struct {
+	// Listeners returns the streaming service's listeners for the BaseApp to register
+	Listeners() map[types.StoreKey][]store.WriteListener
+	// ABCIListener interface for hooking into the ABCI messages from inside the BaseApp
+	ABCIListener ABCIListener
+	// StopNodeOnErr stops the node when true
+	StopNodeOnErr bool
 }
 ```
 
-#### BaseApp registration
+We will also introduce a new `KVStoreWriter` struct as the writer for use in `StoreKVPairWriteListener`. `KVStoreWriter` will expose `ABCIListener` for streaming state change events via `ABCIListener.ListenStoreKVPair`, a `BlockHeight` func for determining the block height for which the state change events belong to and a boolean flag for determine whether the node needs to be stopped or not.
+
+```go
+// KVStoreWriter is used so that we do not need to update the underlying
+// io.Writer inside the StoreKVPairWriteListener everytime we begin writing
+type KVStoreWriter struct {
+	BlockHeight   func() int64
+	listener      ABCIListener
+	stopNodeOnErr bool
+}
+
+// NewKVStoreWriter create an instance of a KVStoreWriter that sends StoreKVPair data to listening service
+func NewKVStoreWriter(
+	listener ABCIListener,
+	stopNodeOnErr bool,
+	blockHeight func() int64,
+) *KVStoreWriter {
+	return &KVStoreWriter{
+		listener:      listener,
+		stopNodeOnErr: stopNodeOnErr,
+		BlockHeight:   blockHeight,
+	}
+}
+
+// Write satisfies io.Writer
+func (iw *KVStoreWriter) Write(b []byte) (int, error) {
+	blockHeight := iw.BlockHeight()
+	if err := iw.listener.ListenStoreKVPair(blockHeight, b); err != nil {
+		if iw.stopNodeOnErr {
+			panic(err)
+		}
+		return 0, err
+	}
+	return len(b), nil
+}
+```
+
+We will expose a `RegisterStreamingService` function for the App to register streaming services.
+```go
+// RegisterStreamingService registers the ABCI streaming service provided by the streaming plugin.
+func RegisterStreamingService(
+	bApp *BaseApp,
+	appOpts servertypes.AppOptions,
+	kodec codec.BinaryCodec,
+	keys map[string]*types.KVStoreKey,
+	streamingService interface{},
```

Comment:
+1 same question.

Reply:
The plugin system is designed to serve plugins of different types. My thinking was that by keeping it generic we could handle the different plugin registrations in `RegisterStreamingService`.

Perhaps it would better to align with the plugin system as follows?

```toml
[streaming]
 
# ADR-038 plugin (ABCIListener)
[streaming.state-listening]
enable = true
keys = []
plugin = "abci_v1"


# Some other part of the system
[streaming.other1]
enable = true
plugin = "name_v1"

# Some other part of the system
[streaming.other2]
enable = true
plugin = "name_v1"
```

Reply:
I'm not following how this would work as written, since all the outputs from `streamingService` are after the type cast to `ABCIListener`.  Are we thinking the addition of more streaming service types would result in a dynamic dispatch inside `RegisterStreamingService` to different (or updated) downstream types?